File: 백준/Gold/16985. Maaaaaaaaaze/Maaaaaaaaaze.java
import java.io.*;
import java.util.*;

public class Main {
    static int[][][] mask = new int[5][5][5]; // [층][가로][세로]
    static int[] batchPos = new int[5];
    static boolean[] batchUsed = new boolean[5];
    final static int INF = 0x3f3f3f3f;
    static int[] dx = { 1, 0, -1, 0, 0, 0 };
    static int[] dy = { 0, 1, 0, -1, 0, 0 };
    static int[] dz = { 0, 0, 0, 0, 1, -1 };

    static List<Pos> starts = new ArrayList<>();
    static int min = INF;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                StringTokenizer st = new StringTokenizer(br.readLine());
                for (int k = 0; k < 5; k++) {
                    mask[i][j][k] = Integer.parseInt(st.nextToken());
                }
            }
        }

        starts.add(new Pos(0, 0, 0));
        starts.add(new Pos(0, 4, 0));
        starts.add(new Pos(4, 0, 0));
        starts.add(new Pos(4, 4, 0));
        starts.add(new Pos(0, 0, 4));
        starts.add(new Pos(0, 4, 4));
        starts.add(new Pos(4, 0, 4));
        starts.add(new Pos(4, 4, 4));

        rotate(0);

        System.out.print(min == INF ? -1 : min);
    }

    public static void batchAndSolve() {
        int[][][] maze = new int[5][5][5];
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                for (int k = 0; k < 5; k++) {
                    maze[i][j][k] = mask[batchPos[i]][j][k];
                }
            }
        }

        for (Pos st : starts) {
            if (maze[st.x][st.y][st.z] == 0) continue;
            Pos dest = new Pos(4 - st.x, 4 - st.y, 4 - st.z);
            if (maze[dest.x][dest.y][dest.z] == 0) continue;

            int lb = Math.abs(st.x - dest.x) + Math.abs(st.y - dest.y) + Math.abs(st.z - dest.z);
            if (lb >= min) continue;

            ArrayDeque<Node> q = new ArrayDeque<>();
            boolean[][][] visited = new boolean[5][5][5];
            visited[st.x][st.y][st.z] = true;
            q.add(new Node(st.x, st.y, st.z, 0));

            while (!q.isEmpty()) {
                Node cur = q.poll();
                if (cur.d >= min) continue;
                if (cur.x == dest.x && cur.y == dest.y && cur.z == dest.z) {
                    min = Math.min(min, cur.d);
                    break;
                }
                for (int dir = 0; dir < 6; dir++) {
                    int nx = cur.x + dx[dir];
                    int ny = cur.y + dy[dir];
                    int nz = cur.z + dz[dir];
                    if (nx < 0 || nx >= 5 || ny < 0 || ny >= 5 || nz < 0 || nz >= 5) continue;
                    if (visited[nx][ny][nz]) continue;
                    if (maze[nx][ny][nz] == 0) continue;
                    visited[nx][ny][nz] = true;
                    q.add(new Node(nx, ny, nz, cur.d + 1));
                }
            }
            if (min == 12) return;
        }
    }

    public static void batch(int cur) {
        if (cur == 5) {
            batchAndSolve();
            return;
        }
        for (int i = 0; i < 5; i++) {
            if (batchUsed[i]) continue;
            batchPos[cur] = i;
            batchUsed[i] = true;
            batch(cur + 1);
            batchUsed[i] = false;
            if (min == 12) return;
        }
    }

    public static void rotate(int cur) {
        if (cur == 5) {
            Arrays.fill(batchUsed, false);
            batch(0);
            return;
        }

        for (int rot = 0; rot < 4; rot++) {
            if (rot > 0) rotateLayer90(cur);
            rotate(cur + 1);
            if (min == 12) return;
        }

    }

    public static void rotateLayer90(int layer) {
        int[][] tmp = new int[5][5];
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                tmp[j][4 - i] = mask[layer][i][j];
            }
        }
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                mask[layer][i][j] = tmp[i][j];
            }
        }
    }

    static class Node {
        int x, y, z, d;
        Node(int x, int y, int z, int d) { this.x = x; this.y = y; this.z = z; this.d = d; }
    }

    static class Pos {
        int x, y, z;
        Pos(int x, int y, int z) { this.x = x; this.y = y; this.z = z; }
    }
}

File: 백준/Gold/16985. Maaaaaaaaaze/Maaaaaaaaaze.kt
/**
 * 5×5 크기의 판이 5개 
 * 하얀 칸은 참가자가 들어갈 수 있는 칸, 검은 칸은 참가자가 들어갈 수 없는 칸
 * 주어진 판들을 시계 방향, 혹은 반시계 방향으로 자유롭게 회전 가능
 * 회전을 완료하고 5개의 판을 원하는 순서대로 배치 가능
 * 정육면체에서 출발점 임의로 선택(꼭짓점 중에서), 도착지는 정반대
 * 
 * 이동 규칙
 * - 면으로 인접한 칸이 참가자가 들어갈 수 있는 칸인 경우
 * 
 * 본인이 설계한 미로 중 가장 적은 이동 횟수로 탈출한 사람이 우승
 * 미로의 입구 혹은 출구가 막혀있거나, 입구에서 출구에 도달할 수 있는 방법이 존재하지 않을 경우에는 탈출 불가능
 * 첫째 줄에 주어진 판으로 설계된 미로를 탈출하는 가장 적은 이동 횟수
 */

import java.util.*

const val MAX = 5
const val INF = 1000


fun rotate(board: Array<IntArray>): Array<IntArray> = 
    Array(MAX) { i -> IntArray(MAX) { j -> board[j][MAX - i - 1]}}

fun main() = with(System.`in`.bufferedReader()) {
    val dx = listOf(1, 0, -1, 0, 0, 0)
    val dy = listOf(0, 1, 0, -1, 0, 0)
    val dz = listOf(0, 0, 0, 0, 1, -1)

    val starts = listOf(
        Triple(0, 0, 0),
        Triple(0, MAX - 1, 0),
        Triple(MAX - 1, 0, 0),
        Triple(MAX - 1, MAX - 1, 0)
    )

    val ends = listOf(
        Triple(MAX - 1, MAX - 1, MAX - 1),
        Triple(MAX - 1, 0, MAX - 1),
        Triple(0, MAX - 1, MAX - 1),
        Triple(0, 0, MAX - 1)
    )

    var boards = Array(MAX) { Array(MAX) { StringTokenizer(readLine()).let { st -> IntArray(MAX) { st.nextToken().toInt() } } } }
    val rotations = Array(MAX) { i ->
        Array(4) { r ->
            var b = boards[i]
            repeat(r) { b = rotate(b) }
            b
        }
    }

    var cube = Array(MAX) { Array(MAX) { IntArray(MAX) } }
    val used = BooleanArray(MAX)
    var min = INF

    fun bfs(id: Int): Int {
        val s = starts[id]
        val e = ends[id]
        if (cube[s.third][s.first][s.second] == 0) return INF
        if (cube[e.third][e.first][e.second] == 0) return INF

        val queue = ArrayDeque<Triple<Int, Int, Int>>()
        val visited = Array(MAX) { Array(MAX) { BooleanArray(MAX) } }
        val dists = Array(MAX) { Array(MAX) { IntArray(MAX) { INF } } }
        visited[s.third][s.first][s.second] = true
        dists[s.third][s.first][s.second] = 0
        queue.offer(s)

        while (!queue.isEmpty()) {
            val c = queue.poll()
            if (c == e) return dists[c.third][c.first][c.second]

            for (i in 0 until 6) {
                val nx = c.first + dx[i]
                val ny = c.second + dy[i]
                val nz = c.third + dz[i]

                if (nx < 0 || ny < 0 || nz < 0 || nx >= MAX || ny >= MAX || nz >= MAX) continue
                if (cube[nz][nx][ny] == 0) continue
                if (visited[nz][nx][ny]) continue

                visited[nz][nx][ny] = true
                dists[nz][nx][ny] = dists[c.third][c.first][c.second] + 1
                queue.offer(Triple(nx, ny, nz))
            }
        }

        return INF
    }

    fun batch(z: Int) {
        if (z == MAX) {
            val result = bfs(0)
            min = minOf(min, result)
            return
        }

        for (i in 0 until MAX) {
            if (used[i]) continue
            cube[z] = boards[i]
            used[i] = true
            batch(z + 1)
            used[i] = false
        }
    }

    fun solve(boardIdx: Int) {
        if (boardIdx == MAX) {
            batch(0)
            return
        }

        for (r in 0 until 4) {
            boards[boardIdx] = rotations[boardIdx][r]
            solve(boardIdx + 1)
        }
    }

    solve(0)
    print(if (min == INF) -1 else min)
}